smart_open/gcs.py

```diff
@@ -470,52 +479,136 @@ def terminate(self):
+                "status code: %i "
+                "response text=%s), "
                 "part #%i, "
-                "%i bytes (total %.3fGB)"
+                "%i bytes (total %.3fGB), "
+                "headers %r"
             ) % (
                 response.status_code,
                 response.text,
                 part_num,
-                self._current_part.tell(),
+                content_length,
                 self._total_size / 1024.0 ** 3,
+                headers,
             )
             raise UploadFailedError(msg, response.status_code, response.text)
         logger.debug("upload of part #%i finished" % part_num)
 
         self._total_parts += 1
+        self._bytes_uploaded += content_length
+        # handle the leftovers
+        self._current_part = io.BytesIO(self._current_part.read())
+        self._current_part.seek(0, io.SEEK_END)
+
+    def _upload_final_part(self):
+        part_num = self._total_parts + 1
+        part_size = self._current_part_size
+        stop = self._total_size - 1
+        start = self._bytes_uploaded
+
+        headers = {
+            'Content-Length': str(part_size),
+            'Content-Range': _make_range_string(start, stop, self._total_size),
+        }
+
+        logger.info(
+          "uploading part #%i, "
+          "%i bytes (total %.3fGB)"
+          "headers %r",
+          part_num,
+          part_size,
+          self._total_size / 1024.0 ** 3,
+          headers,
+        )
+
+        self._current_part.seek(0)
+
+        response = self._session.put(
+            self._resumable_upload_url,
+            data=self._current_part,
+            headers=headers,
+        )
+
+        if response.status_code not in _UPLOAD_COMPLETE_STATUS_CODES:
+            msg = (
+                "upload failed ("
+                "status code: %i ,"
+                "response text=%s), "
+                "part #%i, "
+                "%i bytes (total %.3fGB), "
+                "headers %r"
+            ) % (
+                response.status_code,
+                response.text,
+                part_num,
+                part_size,
+                self._total_size / 1024.0 ** 3,
+                headers,
+              )
+            raise UploadFailedError(msg, response.status_code, response.text)
+        logger.debug("upload of part #%i finished" % part_num)
+
+        self._total_parts += 1
+        self._bytes_uploaded += part_size
         self._current_part = io.BytesIO()
 
     def _upload_empty_part(self):
         logger.debug("creating empty file")
         headers = {'Content-Length': '0'}
         response = self._session.put(self._resumable_upload_url, headers=headers)
-        assert response.status_code in _SUCCESSFUL_STATUS_CODES
+        assert response.status_code in _UPLOAD_COMPLETE_STATUS_CODES
 
         self._total_parts += 1
 
+    @property
+    def _current_part_size(self):
+        return self._current_part.seek(0, io.SEEK_END)
```

Comment:
This property has a nasty side effect - it seeks to the end of the current part, and this side effect is completely obscured by its name. This could cause problems in the future, for example, if someone uses this property in log statements or otherwise.

I think the way forward is to make this a regular method (or even a pure function, as it doesn't really use the full object) and rename this as `_finalize_current_part` to make it obvious that it modifies the buffer significantly.

You could also undo the side effect before returning, but I think the above alternative is better.